### PR TITLE
CA-213224: Avoid concurrent internal bridge creation

### PR DESCRIPTION
**Pull request by robhoes**
When two threads try to bring up the same internal bridge (which can happen if
we boot two VMs attached to the same internal network in parallel), there is a
chance of confliction as bridge creation steps are not atomic.

Signed-off-by: Zheng Li <dev@zheng.li>

Backported and conflicts resolved by Rob Hoes <rob.hoes@citrix.com>, from https://github.com/xapi-project/xen-api/commit/12e387c5f9acb42bba1606cd56e26dca6499e0da.

### CODE REVIEW

File: ocaml/xapi/xapi_network.ml [hunk 1]
```diff
@@ -20,13 +20,24 @@ open D
...
+		debug "Internal bridge %s exists" bridge
+	else
+		(* Atomic test-and-set process *)
+		Mutex.execute internal_bridge_m (fun () ->
+			let current = Net.Bridge.get_all dbg () in
+			if not(List.mem bridge current) then begin
+				let other_config = ["network-uuids", uuid] in
+				debug "Creating internal bridge %s (uuid:%s)" bridge uuid;
+				Net.Bridge.create dbg ~name:bridge ~other_config ();
+			end
```
Comment (mseri) [inline ocaml/xapi/xapi_network.ml:39]: Would it be worth to add a debug string in a inner `else` case?

**Review by lindig — APPROVED** [review]
This backport looks faithful to the original at https://github.com/xapi-project/xen-api/commit/12e387c5f9acb42bba1606cd56e26dca6499e0da

**Review by mseri — APPROVED** [review]
Looks like a faithful backport

File: ocaml/xapi/xapi_network.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi/xapi_network.ml:39]: Perhaps, but I didn't want to change the original patch too much.

**Pull request merged**